### PR TITLE
add ability to taper analysis perts near top of model

### LINKED ISSUE #728: Add ability to taper analysis perturbations near top of model
ensemble spread tends to get very large near the top of the GFS when cycling the EnKF.  To control this growth of spread near the top, the ability to taper (or 'deflate') EnKF analysis perturbations will be added.

### PR DESCRIPTION
**Pull request by jswhit**
Fixes #728

Addresses issue #728 by adding a logical `taperanalperts` to the namelist.  The vertical profile of the taper profile

```
log(ak(k) - ak_top)/log(ak_bot - ak_top)
```

 is controlled by parameters `taperanalperts_aktop` and `taperanalperts_akbot`. Default values are 500 and -1, which results in  a taper that starts at level 19 and has a value of 0.11 at level 1 for the 127 level GFS.

@CatherineThomas-NOAA - I decided it would be simpler and less intrusive to the workflow to include this functionality in the EnKF code instead of the recentering utility.

TODO: implement for regional models
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by RussTreadon-NOAA** [comment]
@jswhit , who might we ask to review these changes?  We need a minimum of two peer reviewers and approvals.

**Comment by jswhit2** [comment]
> @jswhit , who might we ask to review these changes? We need a minimum of two peer reviewers and approvals.

I suggest @CatherineThomas-NOAA and @TingLei-NOAA

**Comment by RussTreadon-NOAA** [comment]
Thanks @jswhit .  Cathy and Ting have been added as peer reviewers.

**Comment by TingLei-NOAA** [comment]
@jswhit  I am a little confused on the profile generated by 
```
log(ak(k) - ak_top)/log(ak_bot - ak_top)
```
For example, as ak(k) approaches ak_top from the positive side (always such case), with a difference from ak_top less than 1, the magnitude of log(ak(k)-ak_top) would increase,  while it is a negative number.  Such values are not what we expect for the tamper coefficients, right?   While  "(log(ak(k))-log(ak_top))/(log(ak_bot/ak_top) seems to be no such issues with 0  on ak_top.    Wondering if you would have better profiles. Please let me know if I missed any implicit properties for ak. Your further clarification will be appreciated.

**Comment by jswhit2** [comment]
@Ting.Lee the default values of akbot and aktop are 500 and  -1.  log(akbot-aktop) = 6.21.  For 500<ak<-1, log(ak-aktop) is decreases from 6.21 to about 7 for ak=1 (the top level of the model).  Thus the taper profile decreases from 1 when ak=500 to about 0.1 when ak=1.

**Comment by TingLei-NOAA** [comment]
@jswhit  Thanks for this explanation.  With the aktop=-1 , this indeed works as your described.

File: src/enkf/gridinfo_nmmb.f90
Comment (TingLei-NOAA) [inline src/enkf/gridinfo_nmmb.f90:129]: Since taperanalperts =.true. won't have actual impacts for gridinfo_wrf.f90 and gridinfo_nmmb.f90 (taper_vert always=1) . Shall we add a test in those program, when taperanalperts=.true.,  the program would abort with error message to avoid any mis-use of taperanalperts=.true. (while the users think they use while not) for those regional applications.

Comment (jswhit2) [reply src/enkf/gridinfo_nmmb.f90:129]: It appears that the wrf and nmmb versions of the code have bitrot - they don't compile (due to missing parallel read/write functions).  Not sure if they are maintained anymore.

**Comment by RussTreadon-NOAA** [comment]
**Orion ctests**

Clone `jswhit:taperanalperts` on Orion and merge `develop` at [b53740a](https://github.com/NOAA-EMC/GSI/commit/b53740a7bd1cc416f634589075b8c8b89f0ef761) into working copy.  Run ctests with following results
```
Test project /work2/noaa/da/rtreadon/git/gsi/pr729/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  786.10 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  844.88 sec
3/7 Test #7: global_enkf ......................   Passed  909.35 sec
4/7 Test #2: rtma .............................   Passed  1269.14 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1404.80 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1641.61 sec
7/7 Test #1: global_4denvar ...................   Passed  2043.23 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 2043.25 sec
```
All tests pass.  This is expected.. This PR only changes files in `src/enkf`.  All _updat_ and _contrl_ `gsi.x` based tests should be identical since the source code is identical.  We also do not expect any difference in the `enkf.x` test.  The taper option added by the PR is off by default.  `src/enkf/params.f90` contains
```
logical, public :: taperanalperts = .false.
```
Thus, the `global_enkf` analysis results should be identical.  The Orion tests confirm this.

**Comment by RussTreadon-NOAA** [comment]
**Hercules ctests**

Repeat Orion test on Hercules with the following results
```
Test project /work/noaa/da/rtreadon/git/gsi/pr729/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.88 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  484.93 sec
3/7 Test #7: global_enkf ......................   Passed  725.90 sec
4/7 Test #2: rtma .............................   Passed  964.86 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1090.78 sec
6/7 Test #5: hafs_4denvar_glbens ..............***Failed  1095.93 sec
7/7 Test #1: global_4denvar ...................   Passed  1741.54 sec

86% tests passed, 1 tests failed out of 7

Total Test time (real) = 1741.64 sec

The following tests FAILED:
          5 - hafs_4denvar_glbens (Failed)
```
`hafs_4denvar_glbens` failed due to
```
The fv3_dynvars are reproducible
The fv3_sfcdata are reproducible
The results between the two runs (hafs_4denvar_glbens_loproc_updat and hafs_4denvar_glbens_loproc_contrl) are not reproducible. Thus, the case has Failed siganl of the regression tests.
```
Comparison of the _loproc_contrl_ and _loproc_updat_ `fv3_tracer` output files shows very small differences in the `o3mr` field
```
o3mr min/max 1=1.4650413e-08,1.6894655e-05 min/max 2=1.4650413e-08,1.6894655e-05 max abs diff=0.0000000450
```
Interestingly, the _hiproc_updat_ `fv3_tracer` file is identical to the `fv3_tracer` file generated by the _loproc_contrl_ and _hiproc_contrl_ runs.   The _loproc_updat_ is the anomalous run.  Non-reproducibility in regional analysis files is a known problem on Hercules.

All other ctests _Passed_ on Hercules.

**Review by TingLei-NOAA — APPROVED** [review]

**Comment by RussTreadon-NOAA** [comment]
**Hera ctests**

Repeat Hercules test on Hera with the following results
```
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr729/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  550.06 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  608.17 sec
3/7 Test #7: global_enkf ......................***Failed  998.36 sec
4/7 Test #2: rtma .............................   Passed  1149.47 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1223.71 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1348.94 sec
7/7 Test #1: global_4denvar ...................   Passed  1986.77 sec

86% tests passed, 1 tests failed out of 7

Total Test time (real) = 1986.81 sec

The following tests FAILED:
          7 - global_enkf (Failed)
```
`global_enkf` failed due to the time threshold test.
```
The runtime for global_enkf_loproc_updat is 168.053597 seconds.  This has exceeded maximum allowable threshold time of 163.541045 seconds, resulting in Failure timethresh of the regression test.
```
A check of `enkf.x` wall times returns
```
global_enkf_hiproc_contrl/stdout:The total amount of wall time                        = 125.916447
global_enkf_hiproc_updat/stdout:The total amount of wall time                        = 128.483509
global_enkf_loproc_contrl/stdout:The total amount of wall time                        = 149.715914
global_enkf_loproc_updat/stdout:The total amount of wall time                        = 170.181588
```
While the _loproc_updat_ `enkf.x` wall time is higher than the _loproc_contrl_ is it not anomalously larger.  This is not a fatal fail.

**Comment by RussTreadon-NOAA** [comment]
**WCOSS2 (Dogwood) ctests**

Repeat Hera test on Dogwood with the following results
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr729/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  604.61 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  665.51 sec
3/7 Test #7: global_enkf ......................   Passed  970.00 sec
4/7 Test #6: hafs_3denvar_hybens ..............   Passed  1212.18 sec
5/7 Test #2: rtma .............................   Passed  1327.94 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1392.89 sec
7/7 Test #1: global_4denvar ...................   Passed  1862.29 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1862.30 sec
```
As expected, all tests pass.

**Comment by RussTreadon-NOAA** [comment]
@jswhit , when you have time, please merge recent commits to the authoritative `develop` into `jswhit:taperanalperts`.  Your branch is currently 3 commits behind `develop`.

I am on leave 4/1 through 4/11.  Hence my running ctests with the hope of merging this PR into `develop` before COB 3/29 pending approval from peer reviewers.

**Comment by jswhit2** [comment]
> @jswhit , when you have time, please merge recent commits to the authoritative `develop` into `jswhit:taperanalperts`. Your branch is currently 3 commits behind `develop`.
> 
> I am on leave 4/1 through 4/11. Hence my running ctests with the hope of merging this PR into `develop` before COB 3/29 pending approval from peer reviewers.

done!

File: src/enkf/gridinfo_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_fv3reg.f90]: Nitpicks
- replace `.eq.` with `==`
- replace `0.` with `0._r_single`
- checking for equality between floating point values should be avoided if possible

File: src/enkf/gridinfo_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_fv3reg.f90]: Replace `0.` with `0._r_single`

File: src/enkf/gridinfo_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_fv3reg.f90:188]: Note:  `one` is declared in `src/gsi/constants.f90` as a type `r_kind` parameter.  `taper_vert` is declared as type `r_single`.  Might consider replacing `one` with `1.0_r_single`.  This is another nitpick comment (ie, not major).

File: src/enkf/gridinfo_gfs.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_gfs.f90:178]: `one` is type `r_kind`.  `taper_vert` is type `r_single`.  Inconsistent precision.

File: src/enkf/gridinfo_gfs.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_gfs.f90]: Same comments as in `gridinfo_fv3reg.f90`

File: src/enkf/gridinfo_gfs.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_gfs.f90]: Same comments as in `gridinfo_fv3reg.f90`

File: src/enkf/gridinfo_nmmb.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_nmmb.f90:130]: `taper_vert` is type `r_single` whereas `one` is type `r_kind`

File: src/enkf/gridinfo_wrf.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridinfo_wrf.f90:218]: Difference precisions.  `taper_vert` is `r_single`.  `one` is `r_kind`

File: src/enkf/inflation.f90
Comment (RussTreadon-NOAA) [inline src/enkf/inflation.f90]: Consider replacing `.eq.` with `==`

File: src/enkf/params.f90
Comment (RussTreadon-NOAA) [inline src/enkf/params.f90]: Should specify the intended precision:  `r_single`, `r_kind`, or ...?

File: src/enkf/params.f90
Comment (RussTreadon-NOAA) [inline src/enkf/params.f90]: Should specify the intended precision:  `r_single`, `r_kind`, or ...?

**Review by RussTreadon-NOAA — COMMENTED** [review]
Changes look good.  Review comments are minor and fall into the _nitpick_ category - nothing serious or major.

**Comment by CatherineThomas-NOAA** [comment]
@jswhit Just want to confirm that the description:

> is controlled by parameters taperanalperts_aktop and taperanalperts_akbot. Default values are 500 and -1, 

is swapped, correct?  aktop=-1 and ak_bot=500?  This is what's in the code and what mathematically makes sense.

**Comment by RussTreadon-NOAA** [comment]
Rerun ctests on Hera Rocky 8 nodes using `taperanalperts` at 87af1e2cd.  All tests pass on rerun.
```
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr729b/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  1330.32 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  1387.52 sec
3/7 Test #7: global_enkf ......................   Passed  1522.12 sec
4/7 Test #6: hafs_3denvar_hybens ..............   Passed  1821.52 sec
5/7 Test #2: rtma .............................   Passed  2052.07 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  2068.81 sec
7/7 Test #1: global_4denvar ...................   Passed  3967.34 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 3967.40 sec
```

**Comment by jswhit2** [comment]
> @jswhit Just want to confirm that the description:
> 
> > is controlled by parameters taperanalperts_aktop and taperanalperts_akbot. Default values are 500 and -1,
> 
> is swapped, correct? aktop=-1 and ak_bot=500? This is what's in the code and what mathematically makes sense.

yes, it is swapped - sorry about that

File: src/enkf/gridinfo_gfs.f90
Comment (jswhit2) [reply src/enkf/gridinfo_gfs.f90]: done

File: src/enkf/gridinfo_gfs.f90
Comment (jswhit2) [reply src/enkf/gridinfo_gfs.f90]: done

File: src/enkf/gridinfo_nmmb.f90
Comment (jswhit2) [reply src/enkf/gridinfo_nmmb.f90:130]: done

File: src/enkf/gridinfo_wrf.f90
Comment (jswhit2) [reply src/enkf/gridinfo_wrf.f90:218]: done

File: src/enkf/inflation.f90
Comment (jswhit2) [reply src/enkf/inflation.f90]: done

File: src/enkf/params.f90
Comment (jswhit2) [reply src/enkf/params.f90]: done by making r_kind

File: src/enkf/params.f90
Comment (jswhit2) [reply src/enkf/params.f90]: done

File: src/enkf/params.f90
Comment (RussTreadon-NOAA) [inline src/enkf/params.f90]: Replace `500.` with `500.0_r_kind` and `-1.` with `-1.0_r_kind` on the line below.

**Review by RussTreadon-NOAA — COMMENTED** [review]
Thank you @jswhit for patiently responding to my nitpicks.  Looks good.  Just two more lines which need precision specified in `params.f90`.  I missed this on my first sweep.  Sorry.

File: src/enkf/params.f90
Comment (jswhit2) [reply src/enkf/params.f90]: done!

**Review by RussTreadon-NOAA — APPROVED** [review]
Great!  Thank you @jswhit for the quick commit.

Approve.

**Review by CatherineThomas-NOAA — APPROVED** [review]
Changes look good to me.  Thanks @jswhit for adding this!

**Pull request merged**
